src/main/scala/beam/utils/mapsapi/googleapi/GoogleAdapter.scala

```diff
@@ -0,0 +1,147 @@
+  }
+
+  private def call(url: String): Future[JsObject] = {
+    val httpRequest = HttpRequest(uri = url)
+    val responseFuture: Future[HttpResponse] = Http().singleRequest(httpRequest)
+    responseFuture.map { response =>
+      val inputStream = response.entity.dataBytes.runWith(StreamConverters.asInputStream(timeout))
+      Json.parse(inputStream).as[JsObject]
+    }
+  }
+
+  private def parseRoutes(jsRoutes: Seq[JsValue]): Seq[Route] = {
+    jsRoutes.map { route =>
+      val firstAndUniqueLeg = (route \ "legs").as[JsArray].value.head
+      parseRoute(firstAndUniqueLeg.as[JsObject])
+    }
+  }
+
+  private def writeToFileIfSetup(jsObject: JsObject): JsObject = {
+    if (outputResponseToFile.isDefined) {
+      FileUtils.writeStringToFile(outputResponseToFile.get.toFile, Json.prettyPrint(jsObject), StandardCharsets.UTF_8)
+    }
+    jsObject
+  }
+
+  private def toRoutes(jsObject: JsObject): Seq[Route] = {
+
+    parseRoutes((jsObject \ "routes").as[JsArray].value)
+  }
+
+  private def parseRoute(jsObject: JsObject): Route = {
+    val segments = parseSegments((jsObject \ "steps").as[JsArray].value)
+    val distanceInMeter = (jsObject \ "distance" \ "value").as[Int]
+    val durationInSeconds = (jsObject \ "duration" \ "value").as[Int]
+    val startLocation = parseWgsCoordinate(jsObject \ "start_location")
+    val endLocation = parseWgsCoordinate(jsObject \ "end_location")
+    Route(startLocation, endLocation, distanceInMeter, durationInSeconds, segments)
+  }
+
+  private def parseStep(jsObject: JsObject): Segment = {
+    Segment(
+      coordinates = PolyDecoder.decode((jsObject \ "polyline" \ "points").as[String]),
+      lengthInMeters = (jsObject \ "distance" \ "value").as[Int],
+      durationInSeconds = Some((jsObject \ "duration" \ "value").as[Int])
+    )
+  }
+
+  private def parseSegments(steps: Seq[JsValue]): Seq[Segment] = {
+    steps.map { step =>
+      parseStep(step.as[JsObject])
+    }
+  }
+
+  private def parseWgsCoordinate(position: JsLookupResult) = {
+    WgsCoordinate(
+      latitude = (position \ "lat").as[Double],
+      longitude = (position \ "lng").as[Double]
+    )
+  }
+
+  override def close(): Unit = {
+    Http().shutdownAllConnectionPools
+      .andThen {
+        case _ =>
+          if (!materializer.isShutdown) materializer.shutdown()
+          system.terminate()
+      }
+  }
+
+}
+
+object GoogleAdapter {
+
+  private[googleapi] def buildUrl(
+    apiKey: String,
+    origin: WgsCoordinate,
+    destination: WgsCoordinate,
+    departureAt: LocalDateTime,
+    mode: TravelModes.TravelMode,
+    constraints: Set[TravelConstraints.TravelConstraint]
+  ): String = {
+    // avoid=tolls|highways|ferries
+    val originStr = s"${origin.latitude},${origin.longitude}"
+    val destinationStr = s"${destination.latitude},${destination.longitude}"
+    val params = Seq(
+      "mode=driving",
+      "language=en",
+      "units=metric",
+      "alternatives=true",
```

Comment:
can we switch algorithm for request, e.g. optimistic/pessimistic/best guess

Reply:
done here:
https://github.com/LBNL-UCB-STI/beam/pull/2667/files#diff-919577ab6fbd86949a2146cf6cd8b35cR133